File: ios/ClosetCraftAR/ARMeasureModule.swift
/**
 * ARMeasureModule.swift
 * Native ARKit measurement module for ClosetCraft
 *
 * Provides real-time spatial wall measurement using ARKit's world tracking.
 * Communicates with React Native via the bridge (ARMeasureModule.m).
 *
 * Accuracy: ±0.5 inches on supported devices (iPhone 6s+, A9 chip)
 */

import ARKit
import SceneKit
import React

@objc(ARMeasureModule)
class ARMeasureModule: RCTEventEmitter {
    
    // MARK: - Properties
    
    private var arView: ARSCNView?
    private var arSession: ARSession?
    private var startPoint: SCNVector3?
    private var measureNodes: [SCNNode] = []
    private var lineNode: SCNNode?
    private var labelNode: SCNNode?
    private var isSessionRunning = false
    
    // Calibration state
    private var calibrationMeasurements: [Float] = []
    private var calibrationComplete = false
    
    // MARK: - RCTEventEmitter
    
    override func supportedEvents() -> [String] {
        return [
            "onMeasurement",
            "onARStatus",
            "onPlaneDetected",
            "onTrackingStateChanged",
            "onError"
        ]
    }
    
    override static func requiresMainQueueSetup() -> Bool {
        return true
    }
    
    // MARK: - Device Support Check
    
    @objc func checkSupport(_ resolve: @escaping RCTPromiseResolveBlock,
                            rejecter reject: RCTPromiseRejectBlock) {
        let supported = ARWorldTrackingConfiguration.isSupported
        let supportsSceneReconstruction: Bool
        if #available(iOS 13.4, *) {
            supportsSceneReconstruction = ARWorldTrackingConfiguration.supportsSceneReconstruction(.mesh)
        } else {
            supportsSceneReconstruction = false
        }
        
        resolve([
            "arSupported": supported,
            "lidarAvailable": supportsSceneReconstruction,
            "deviceModel": UIDevice.current.model,
            "systemVersion": UIDevice.current.systemVersion,
        ])
    }
    
    // MARK: - Session Management
    
    @objc func startSession(_ options: NSDictionary,
                            resolve: @escaping RCTPromiseResolveBlock,
                            rejecter reject: @escaping RCTPromiseRejectBlock) {
        guard ARWorldTrackingConfiguration.isSupported else {
            reject("AR_NOT_SUPPORTED", "ARKit is not supported on this device", nil)
            return
        }
        
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            
            let config = ARWorldTrackingConfiguration()
            config.planeDetection = [.vertical, .horizontal]
            config.environmentTexturing = .automatic
            config.isLightEstimationEnabled = true
            
            // Enable LiDAR mesh if available
            if #available(iOS 13.4, *) {
                if ARWorldTrackingConfiguration.supportsSceneReconstruction(.mesh) {
                    config.sceneReconstruction = .mesh
                }
            }
            
            let session = ARSession()
            session.delegate = self
            session.run(config, options: [.resetTracking, .removeExistingAnchors])
            
            self.arSession = session
            self.isSessionRunning = true
            self.startPoint = nil
            self.calibrationComplete = false
            self.calibrationMeasurements.removeAll()
            self.clearMeasureVisuals()
            
            resolve([
                "status": "started",
                "planeDetection": true,
            ])
        }
    }
    
    @objc func stopSession() {
        DispatchQueue.main.async { [weak self] in
            self?.arSession?.pause()
            self?.arSession = nil
            self?.isSessionRunning = false
            self?.startPoint = nil
            self?.clearMeasureVisuals()
        }
    }
    
    // MARK: - Measurement
    
    @objc func placeMeasurePoint(_ screenX: Float, _ screenY: Float,
                                  resolve: @escaping RCTPromiseResolveBlock,
                                  rejecter reject: @escaping RCTPromiseRejectBlock) {
        guard let session = arSession, isSessionRunning else {
            reject("NO_SESSION", "AR session is not active", nil)
            return
        }
        
        guard let frame = session.currentFrame else {
            reject("NO_FRAME", "No AR frame available", nil)
            return
        }
        
        let screenPoint = CGPoint(x: CGFloat(screenX), y: CGFloat(screenY))
        
        // Perform raycast against detected planes
        guard let query = frame.raycastQuery(from: screenPoint,
                                              allowing: .existingPlaneGeometry,
                                              alignment: .any) else {
            reject("RAYCAST_FAILED", "Could not create raycast query", nil)
            return
        }
        
        let results = session.raycast(query)
        
        guard let result = results.first else {
            // Try broader target if plane not detected
            let fallbackQuery = frame.raycastQuery(from: screenPoint,
                                                     allowing: .estimatedPlane,
                                                     alignment: .any)
            if let fallbackQuery = fallbackQuery {
                let fallbackResults = session.raycast(fallbackQuery)
                if let fallbackResult = fallbackResults.first {
                    processHitResult(fallbackResult, resolve: resolve, isEstimated: true)
                    return
                }
            }
            reject("NO_SURFACE", "No surface detected. Move the device slowly to scan the area.", nil)
            return
        }
        
        processHitResult(result, resolve: resolve, isEstimated: false)
    }
    
    private func processHitResult(_ result: ARRaycastResult,
                                   resolve: @escaping RCTPromiseResolveBlock,
                                   isEstimated: Bool) {
        let position = SCNVector3(
            result.worldTransform.columns.3.x,
            result.worldTransform.columns.3.y,
            result.worldTransform.columns.3.z
        )
        
        if let start = startPoint {
            // Calculate distance between two points
            let dx = position.x - start.x
            let dy = position.y - start.y
            let dz = position.z - start.z
            let distanceMeters = sqrt(dx * dx + dy * dy + dz * dz)
            let distanceInches = distanceMeters * 39.3701
            
            // Determine orientation (horizontal = width, vertical = height)
            let horizontalDist = sqrt(dx * dx + dz * dz)
            let verticalDist = abs(dy)
            let orientation = verticalDist > horizontalDist ? "vertical" : "horizontal"
            
            // Add visual line
            DispatchQueue.main.async { [weak self] in
                self?.addMeasureLine(from: start, to: position, distance: distanceInches)
                self?.addMeasureDot(at: position, color: .systemGreen)
            }
            
            startPoint = nil
            
            resolve([
                "status": "measured",
                "distanceInches": round(distanceInches * 10) / 10,  // Round to 0.1"
                "distanceMeters": round(distanceMeters * 1000) / 1000,
                "orientation": orientation,
                "estimated": isEstimated,
                "startPoint": [start.x, start.y, start.z],
                "endPoint": [position.x, position.y, position.z],
                "confidence": isEstimated ? "low" : "high",
            ])
            
            sendEvent(withName: "onMeasurement", body: [
                "distance": round(distanceInches * 10) / 10,
                "orientation": orientation,
            ])
            
        } else {
            // First point — set start
            startPoint = position
            
            DispatchQueue.main.async { [weak self] in
                self?.addMeasureDot(at: position, color: .systemOrange)
            }
            
            resolve([
                "status": "startPointSet",
                "point": [position.x, position.y, position.z],
                "estimated": isEstimated,
            ])
        }
    }
    
    // MARK: - Reset
    
    @objc func resetMeasurement() {
        startPoint = nil
        DispatchQueue.main.async { [weak self] in
            self?.clearMeasureVisuals()
        }
    }
    
    @objc func resetAll() {
        startPoint = nil
        calibrationComplete = false
        calibrationMeasurements.removeAll()
        
        DispatchQueue.main.async { [weak self] in
            self?.clearMeasureVisuals()
        }
        
        // Re-run session
        if let session = arSession {
            let config = ARWorldTrackingConfiguration()
            config.planeDetection = [.vertical, .horizontal]
            config.environmentTexturing = .automatic
            session.run(config, options: [.resetTracking, .removeExistingAnchors])
        }
    }
    
    // MARK: - Visual Helpers
    
    private func addMeasureDot(at position: SCNVector3, color: UIColor) {
        guard let scene = arView?.scene else { return }
        
        let sphere = SCNSphere(radius: 0.008)
        sphere.firstMaterial?.diffuse.contents = color
        sphere.firstMaterial?.lightingModel = .constant
        
        let node = SCNNode(geometry: sphere)
        node.position = position
        scene.rootNode.addChildNode(node)
        measureNodes.append(node)
        
        // Pulse animation
        let pulse = SCNAction.sequence([
            SCNAction.scale(to: 1.3, duration: 0.15),
            SCNAction.scale(to: 1.0, duration: 0.15),
        ])
        node.runAction(pulse)
    }
    
    private func addMeasureLine(from: SCNVector3, to: SCNVector3, distance: Float) {
        guard let scene = arView?.scene else { return }
        
        // Remove previous line
        lineNode?.removeFromParentNode()
        labelNode?.removeFromParentNode()
        
        // Create cylinder between two points
        let dx = to.x - from.x
        let dy = to.y - from.y
        let dz = to.z - from.z
        let length = sqrt(dx*dx + dy*dy + dz*dz)
        
        let cylinder = SCNCylinder(radius: 0.002, height: CGFloat(length))
        cylinder.firstMaterial?.diffuse.contents = UIColor(red: 0.886, green: 0.725, blue: 0.498, alpha: 1.0)
        cylinder.firstMaterial?.lightingModel = .constant
        
        let node = SCNNode(geometry: cylinder)
        node.position = SCNVector3(
            (from.x + to.x) / 2,
            (from.y + to.y) / 2,
            (from.z + to.z) / 2
        )
        
        // Orient cylinder between points
        node.look(at: to, up: scene.rootNode.worldUp, localFront: SCNVector3(0, 1, 0))
        
        scene.rootNode.addChildNode(node)
        lineNode = node
        measureNodes.append(node)
        
        // Add floating distance label
        let ft = Int(distance) / 12
        let inches = Int(distance) % 12
        let labelText = ft > 0 ? "\(ft)'\(inches)\"" : "\(Int(distance))\""
        
        let text = SCNText(string: labelText, extrusionDepth: 0.5)
        text.font = UIFont.systemFont(ofSize: 8, weight: .bold)
        text.firstMaterial?.diffuse.contents = UIColor.white
        text.firstMaterial?.lightingModel = .constant
        text.flatness = 0.1
        
        let textNode = SCNNode(geometry: text)
        textNode.scale = SCNVector3(0.003, 0.003, 0.003)
        textNode.position = SCNVector3(
            (from.x + to.x) / 2,
            (from.y + to.y) / 2 + 0.04,
            (from.z + to.z) / 2
        )
        
        // Billboard constraint so label always faces camera
        let billboard = SCNBillboardConstraint()
        billboard.freeAxes = .all
        textNode.constraints = [billboard]
        
        scene.rootNode.addChildNode(textNode)
        labelNode = textNode
        measureNodes.append(textNode)
    }
    
    private func clearMeasureVisuals() {
        measureNodes.forEach { $0.removeFromParentNode() }
        measureNodes.removeAll()
        lineNode = nil
        labelNode = nil
    }
}

// MARK: - ARSessionDelegate

extension ARMeasureModule: ARSessionDelegate {
    
    func session(_ session: ARSession, didAdd anchors: [ARAnchor]) {
        for anchor in anchors {
            if let planeAnchor = anchor as? ARPlaneAnchor {
                let alignment = planeAnchor.alignment == .vertical ? "vertical" : "horizontal"
                sendEvent(withName: "onPlaneDetected", body: [
                    "id": planeAnchor.identifier.uuidString,
                    "alignment": alignment,
                    "width": planeAnchor.extent.x,
                    "height": planeAnchor.extent.z,
                ])
            }
        }
    }
    
    func session(_ session: ARSession, cameraDidChangeTrackingState camera: ARCamera) {
        var status: String
        var message: String
        
        switch camera.trackingState {
        case .notAvailable:
            status = "notAvailable"
            message = "AR tracking is not available"
        case .limited(let reason):
            status = "limited"
            switch reason {
            case .initializing:
                message = "Initializing — move device slowly"
            case .excessiveMotion:
                message = "Too much motion — hold steady"
            case .insufficientFeatures:
                message = "Not enough features — point at textured surfaces"
            case .relocalizing:
                message = "Relocalizing..."
            @unknown default:
                message = "Limited tracking"
            }
        case .normal:
            status = "normal"
            message = "Tracking is good"
        }
        
        sendEvent(withName: "onTrackingStateChanged", body: [
            "status": status,
            "message": message,
        ])
    }
    
    func session(_ session: ARSession, didFailWithError error: Error) {
        sendEvent(withName: "onError", body: [
            "error": error.localizedDescription,
        ])
    }
}

File: android/app/src/main/java/com/closetcraft/app/ARMeasureModule.kt
/**
 * ARMeasureModule.kt
 * Native ARCore measurement module for ClosetCraft (Android)
 *
 * Provides spatial wall measurement using ARCore's hit testing and plane detection.
 * Communicates with React Native via the bridge (ARMeasurePackage.kt).
 *
 * Accuracy: ±0.5 inches on ARCore-supported devices
 */

package com.closetcraft.app

import android.app.Activity
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import com.google.ar.core.*
import com.google.ar.core.exceptions.*
import kotlin.math.abs
import kotlin.math.roundToInt
import kotlin.math.sqrt

class ARMeasureModule(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    private var arSession: Session? = null
    private var startPoint: FloatArray? = null
    private var isSessionRunning = false

    override fun getName(): String = "ARMeasureModule"

    // ─── Device Support Check ────────────────────────────────────

    @ReactMethod
    fun checkSupport(promise: Promise) {
        try {
            val availability = ArCoreApk.getInstance()
                .checkAvailability(reactApplicationContext)

            val supported = when (availability) {
                ArCoreApk.Availability.SUPPORTED_INSTALLED,
                ArCoreApk.Availability.SUPPORTED_APK_TOO_OLD,
                ArCoreApk.Availability.SUPPORTED_NOT_INSTALLED -> true
                else -> false
            }

            val result = Arguments.createMap().apply {
                putBoolean("arSupported", supported)
                putBoolean("lidarAvailable", false) // Android doesn't have LiDAR (yet)
                putString("deviceModel", android.os.Build.MODEL)
                putString("systemVersion", "Android ${android.os.Build.VERSION.RELEASE}")
                putString("arCoreStatus", availability.name)
            }
            promise.resolve(result)
        } catch (e: Exception) {
            val result = Arguments.createMap().apply {
                putBoolean("arSupported", false)
                putBoolean("lidarAvailable", false)
                putString("error", e.message)
            }
            promise.resolve(result)
        }
    }

    // ─── Session Management ──────────────────────────────────────

    @ReactMethod
    fun startSession(options: ReadableMap, promise: Promise) {
        val activity = currentActivity
        if (activity == null) {
            promise.reject("NO_ACTIVITY", "No active activity found")
            return
        }

        try {
            // Check if ARCore needs to be installed
            when (ArCoreApk.getInstance().requestInstall(activity, true)) {
                ArCoreApk.InstallStatus.INSTALL_REQUESTED -> {
                    promise.reject("INSTALL_REQUESTED",
                        "ARCore installation requested. Please restart the measurement.")
                    return
                }
                ArCoreApk.InstallStatus.INSTALLED -> { /* Continue */ }
            }

            val session = Session(activity)
            val config = Config(session).apply {
                planeFindingMode = Config.PlaneFindingMode.HORIZONTAL_AND_VERTICAL
                updateMode = Config.UpdateMode.LATEST_CAMERA_IMAGE
                focusMode = Config.FocusMode.AUTO
                lightEstimationMode = Config.LightEstimationMode.ENVIRONMENTAL_HDR

                // Enable depth if supported
                if (session.isDepthModeSupported(Config.DepthMode.AUTOMATIC)) {
                    depthMode = Config.DepthMode.AUTOMATIC
                }
            }

            session.configure(config)
            session.resume()

            arSession = session
            isSessionRunning = true
            startPoint = null

            val result = Arguments.createMap().apply {
                putString("status", "started")
                putBoolean("planeDetection", true)
                putBoolean("depthEnabled",
                    config.depthMode == Config.DepthMode.AUTOMATIC)
            }
            promise.resolve(result)

            emitEvent("onARStatus", Arguments.createMap().apply {
                putString("status", "started")
                putString("message", "AR session active. Point at walls to detect surfaces.")
            })

        } catch (e: UnavailableArcoreNotInstalledException) {
            promise.reject("ARCORE_NOT_INSTALLED",
                "ARCore is not installed. Please install it from the Play Store.")
        } catch (e: UnavailableApkTooOldException) {
            promise.reject("ARCORE_OUTDATED",
                "ARCore needs to be updated. Please update from the Play Store.")
        } catch (e: UnavailableSdkTooOldException) {
            promise.reject("SDK_TOO_OLD",
                "This app needs to be updated to support your device's ARCore version.")
        } catch (e: UnavailableDeviceNotCompatibleException) {
            promise.reject("DEVICE_NOT_COMPATIBLE",
                "This device does not support ARCore.")
        } catch (e: Exception) {
            promise.reject("AR_ERROR", "Failed to start AR session: ${e.message}")
        }
    }

    @ReactMethod
    fun stopSession() {
        try {
            arSession?.pause()
            arSession?.close()
        } catch (e: Exception) {
            // Ignore cleanup errors
        }
        arSession = null
        isSessionRunning = false
        startPoint = null
    }

    // ─── Measurement ─────────────────────────────────────────────

    @ReactMethod
    fun placeMeasurePoint(screenX: Float, screenY: Float, promise: Promise) {
        val session = arSession
        if (session == null || !isSessionRunning) {
            promise.reject("NO_SESSION", "AR session is not active")
            return
        }

        try {
            val frame = session.update()

            // Check tracking state
            if (frame.camera.trackingState != TrackingState.TRACKING) {
                emitEvent("onTrackingStateChanged", Arguments.createMap().apply {
                    putString("status", "limited")
                    putString("message", "Tracking lost — move device slowly")
                })
                promise.reject("NOT_TRACKING",
                    "Device is not tracking. Move slowly and point at textured surfaces.")
                return
            }

            // Perform hit test
            val hitResults = frame.hitTest(screenX, screenY)

            // Filter for plane hits first, then estimated planes
            val bestHit = hitResults.firstOrNull { hit ->
                val trackable = hit.trackable
                trackable is Plane &&
                    trackable.isPoseInPolygon(hit.hitPose) &&
                    trackable.trackingState == TrackingState.TRACKING
            } ?: hitResults.firstOrNull { hit ->
                hit.trackable.trackingState == TrackingState.TRACKING
            }

            if (bestHit == null) {
                promise.reject("NO_SURFACE",
                    "No surface detected. Move closer or point at a wall/floor.")
                return
            }

            val pose = bestHit.hitPose
            val point = floatArrayOf(
                pose.tx(), pose.ty(), pose.tz()
            )
            val isEstimated = bestHit.trackable !is Plane

            val currentStart = startPoint
            if (currentStart != null) {
                // Calculate distance
                val dx = point[0] - currentStart[0]
                val dy = point[1] - currentStart[1]
                val dz = point[2] - currentStart[2]
                val distanceMeters = sqrt(
                    (dx * dx + dy * dy + dz * dz).toDouble()
                ).toFloat()
                val distanceInches = distanceMeters * 39.3701f

                // Determine orientation
                val horizontalDist = sqrt((dx * dx + dz * dz).toDouble()).toFloat()
                val verticalDist = abs(dy)
                val orientation = if (verticalDist > horizontalDist) "vertical" else "horizontal"

                startPoint = null

                val result = Arguments.createMap().apply {
                    putString("status", "measured")
                    putDouble("distanceInches",
                        (distanceInches * 10).roundToInt().toDouble() / 10)
                    putDouble("distanceMeters",
                        (distanceMeters * 1000).roundToInt().toDouble() / 1000)
                    putString("orientation", orientation)
                    putBoolean("estimated", isEstimated)
                    putString("confidence", if (isEstimated) "low" else "high")

                    val startArr = Arguments.createArray().apply {
                        pushDouble(currentStart[0].toDouble())
                        pushDouble(currentStart[1].toDouble())
                        pushDouble(currentStart[2].toDouble())
                    }
                    putArray("startPoint", startArr)

                    val endArr = Arguments.createArray().apply {
                        pushDouble(point[0].toDouble())
                        pushDouble(point[1].toDouble())
                        pushDouble(point[2].toDouble())
                    }
                    putArray("endPoint", endArr)
                }
                promise.resolve(result)

                emitEvent("onMeasurement", Arguments.createMap().apply {
                    putDouble("distance",
                        (distanceInches * 10).roundToInt().toDouble() / 10)
                    putString("orientation", orientation)
                })

            } else {
                // Set start point
                startPoint = point

                val result = Arguments.createMap().apply {
                    putString("status", "startPointSet")
                    val pointArr = Arguments.createArray().apply {
                        pushDouble(point[0].toDouble())
                        pushDouble(point[1].toDouble())
                        pushDouble(point[2].toDouble())
                    }
                    putArray("point", pointArr)
                    putBoolean("estimated", isEstimated)
                }
                promise.resolve(result)
            }

            // Report detected planes
            for (plane in session.getAllTrackables(Plane::class.java)) {
                if (plane.trackingState == TrackingState.TRACKING) {
                    val alignment = when (plane.type) {
                        Plane.Type.VERTICAL -> "vertical"
                        Plane.Type.HORIZONTAL_UPWARD_FACING -> "horizontal_up"
                        Plane.Type.HORIZONTAL_DOWNWARD_FACING -> "horizontal_down"
                        else -> "unknown"
                    }
                    emitEvent("onPlaneDetected", Arguments.createMap().apply {
                        putString("alignment", alignment)
                        putDouble("width", plane.extentX.toDouble())
                        putDouble("height", plane.extentZ.toDouble())
                    })
                }
            }

        } catch (e: NotTrackingException) {
            promise.reject("NOT_TRACKING", "Device lost tracking: ${e.message}")
        } catch (e: Exception) {
            promise.reject("MEASURE_ERROR", "Measurement failed: ${e.message}")
        }
    }

    // ─── Reset ───────────────────────────────────────────────────

    @ReactMethod
    fun resetMeasurement() {
        startPoint = null
    }

    @ReactMethod
    fun resetAll() {
        startPoint = null
        val session = arSession ?: return
        try {
            val config = Config(session).apply {
                planeFindingMode = Config.PlaneFindingMode.HORIZONTAL_AND_VERTICAL
                updateMode = Config.UpdateMode.LATEST_CAMERA_IMAGE
            }
            session.configure(config)
            session.resume()
        } catch (e: Exception) {
            emitEvent("onError", Arguments.createMap().apply {
                putString("error", "Reset failed: ${e.message}")
            })
        }
    }

    // ─── Event Helper ────────────────────────────────────────────

    private fun emitEvent(name: String, params: WritableMap) {
        reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(name, params)
    }

    // ─── Lifecycle ───────────────────────────────────────────────

    override fun onCatalystInstanceDestroy() {
        stopSession()
        super.onCatalystInstanceDestroy()
    }
}
